src/controller/tests/TestReadChunking.cpp

```diff
@@ -509,13 +621,240 @@ void TestCommandInteraction::TestDynamicEndpoint(nlTestSuite * apSuite, void * a
     emberAfClearDynamicEndpoint(0);
 }
 
+/*
+ * The tests below are for testing deatiled bwhavior when the attributes are modified between two chunks. In this test, we only care
+ * above whether we will receive correct attribute values in reasonable messages with reduced reporting traffic.
+ */
+void TestCommandInteraction::TestSetDirtyBetweenChunks(nlTestSuite * apSuite, void * apContext)
+{
+    TestContext & ctx                    = *static_cast<TestContext *>(apContext);
+    auto sessionHandle                   = ctx.GetSessionBobToAlice();
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
+
+    // Initialize the ember side server logic
+    InitDataModelHandler(&ctx.GetExchangeManager());
+
+    app::InteractionModelEngine::GetInstance()->GetReportingEngine().SetWriterReserved(0);
+    app::InteractionModelEngine::GetInstance()->GetReportingEngine().SetMaxAttributesPerChunk(2);
+
+    DataVersion dataVersionStorage1[ArraySize(testEndpointClusters)];
+    DataVersion dataVersionStorage5[ArraySize(testEndpoint5Clusters)];
+
+    gMutableAttrAccess.Reset();
+
+    // Register our fake dynamic endpoint.
+    emberAfSetDynamicEndpoint(0, kTestEndpointId, &testEndpoint, 0, 0, Span<DataVersion>(dataVersionStorage1));
+    emberAfSetDynamicEndpoint(1, kTestEndpointId5, &testEndpoint5, 0, 0, Span<DataVersion>(dataVersionStorage5));
+
+// The ReadHandler uses ScheduleWork to kick the report engine without a network traffic, which cannot be handled by DriveIO
+// directly. Also, we want the event loop to stop when we fully received one report. But DrainAndServiceIO cannot be stopped by some
+// conditions.
+#define DRIVE_IO_UNTIL_END_OF_NEXT_REPORT(timeoutSeconds)                                                                          \
+    readCallback.mOnReportEnd = false;                                                                                             \
+    ctx.GetIOContext().DriveIOUntil(System::Clock::Seconds16(timeoutSeconds), [&]() { return readCallback.mOnReportEnd; });        \
+    NL_TEST_ASSERT(apSuite, readCallback.mOnReportEnd);
+
+    {
+        app::AttributePathParams attributePath;
+        app::ReadPrepareParams readParams(sessionHandle);
+
+        readParams.mpAttributePathParamsList    = &attributePath;
+        readParams.mAttributePathParamsListSize = 1;
+        readParams.mMinIntervalFloorSeconds     = 0;
+        readParams.mMaxIntervalCeilingSeconds   = 2;
+
+        //// TEST 1 -- Read using wildcard paths
+        {
```

Comment:
The way these tests are designed makes it quite a bit difficult to decipher what's happening.

I'd very much have liked to see some kind of 'instruction list' like construct, similar to what I had done for `TestBufferedReadCallback.cpp`:

```
enum Attribute {
    kAttributeA,
    kAttributeB,
    kAttributeC
};

struct AttributeVersion {
     Attribute attribute;
     uint16_t version;
};

struct SingleExecutionRound {
    std::vector<Attribute> DirtyInstructions;
    std::vector<AttributeVersion> ReceivedAttributes;
};


ExecuteTest(std::vector<SingleExecutionRound> executionList);
```


Mark A, B dirty, and in the received report, we see A1, B1:
```
ExecuteTest({
                           { {AttributeA, AttributeB},  { {AttributeA, 1}, {AttributeB, 1} } },
                       });
```
Mark A, B, C dirty, and we deliver A, B in the first chunked report, and C later. This test requires artificially limiting the # of attributes that can be sent to 2 in a report:

LimitReportToAttributes(2);
ExecuteTest({
                           { {AttributeA, AttributeB, AttributeC},  { {AttributeA, 1}, {AttributeB, 1} } },
                          {  { },                                                         { {AttributeC, 1} }
                       });
```

etc...

Reply:
Each execution round is a single execution of the engine + receive processing. You can continue to use the 'mark the dirty items in the `OnAttributeData` trick you've employed above.

Reply:
Thanks for advice, infact, wrapping all the works into one function makes the code a bit messy since there are things should be checked in different forms.

Finally ended up with a extra struct, and some template functions / structs / variables to make the code clearer.